File: apps/api/src/services/action_sections.py
"""Sections on an action form (§328; db 0081; `action-types` p.122-124).

    "The action form can be customized with sections. These sections provide a
     logical grouping of parameters to organize an action form. Sections also
     support columns, descriptions, and conditional overrides." (p.122)

    "Sections are also collapsible, can be hidden entirely, and can make use of
     conditional overrides… A section can be hidden at first and only shown
     based on a prior parameter." (p.123)

    "Parameters and sections display in the form based on their order in this
     Form Content section." (p.124)

**A section changes what a form looks like and nothing else.** An action with
every section deleted submits exactly the same values from exactly the same
parameters — which is why `visibility` below returns what to *draw* and never
touches what is bound, checked or written. A hidden section's parameters are
still declared, still defaulted and still validated; p.123 is about a form, not
about a contract.

That distinction is the one thing a reader of this module has to hold on to,
because the alternative is inviting: if a section is hidden, why send its
parameters at all? Because `check_criteria` and `bind_parameters` are what
decides a submission, and a form that quietly dropped values would mean the
same action did different things depending on which boxes were on screen.
"""
from __future__ import annotations

import json
from typing import Any
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncConnection

from ..lib.db import fetch_all, fetch_one
from .actions import _passes  # the condition grammar decision 0007 already has

# **p.123's "one or two columns" is not checked here, and the first draft
# checked it.** `SectionIn.columns` is `Literal[1, 2]` and `action_sections`
# carries `CHECK (columns IN (1, 2))` — two layers that both refuse three,
# either side of this one. The check was unreachable: a mutation sweep deleted
# it and every test stayed green, because nothing can call this with a number
# the request model let through. §213's rule, and the reasoning stands in its
# place rather than the line.


def _json(value: Any) -> Any:
    return json.loads(value) if isinstance(value, str) else value


def visibility(
    section: dict[str, Any],
    *,
    bound: dict[str, Any],
    user: dict[str, Any],
) -> bool:
    """Whether this section is drawn, given what has been filled in so far.

    p.123's three states, in the order they override each other:

    * `hidden` — "can be hidden entirely". Always off, whatever else says.
    * `visible_when` — "hidden at first and only shown based on a prior
      parameter". Evaluated against the values in hand.
    * neither — always on, which is every section anybody made before
      configuring one of the above.

    **A condition that cannot be decided means hidden**, which is the opposite
    of `check_criteria`'s unevaluable-fails-closed only in appearance: both
    refuse to guess. A section shown because its condition was unreadable would
    put parameters in front of somebody on the strength of a broken rule, and
    p.123's whole point is that the form asks for things "under the appropriate
    circumstances".
    """
    if section.get("hidden"):
        return False
    condition = _json(section.get("visible_when"))
    if not condition:
        return True
    try:
        return _passes(condition, bound=bound, user=user)
    except Exception:  # noqa: BLE001 - `_Unevaluable` and anything a bad
        # document can throw. Named broadly on purpose: this is a *display*
        # decision reading a JSON blob somebody typed, and a form that raised
        # a 500 because a section's condition was malformed would take the
        # whole action down for a cosmetic rule.
        return False


async def list_sections(
    conn: AsyncConnection, action_type_id: UUID
) -> list[dict[str, Any]]:
    """p.124's Form Content order, with each section's parameters inside it.

    One statement per kind rather than one per section: a form with six
    sections is six round trips otherwise, and this is read every time an
    action's definition is opened.
    """
    sections = await fetch_all(
        conn,
        """
        SELECT id, title, description, columns, collapsible, collapsed,
               hidden, visible_when, sort_order
          FROM action_sections
         WHERE action_type_id = :atid
         ORDER BY sort_order, title
        """,
        {"atid": str(action_type_id)},
    )
    members = await fetch_all(
        conn,
        """
        SELECT section_id, api_name
          FROM action_parameters
         WHERE action_type_id = :atid AND section_id IS NOT NULL
         ORDER BY sort_order, api_name
        """,
        {"atid": str(action_type_id)},
    )
    inside: dict[str, list[str]] = {}
    for row in members:
        inside.setdefault(str(row["section_id"]), []).append(str(row["api_name"]))
    return [
        {**dict(s), "visible_when": _json(s["visible_when"]),
         "parameters": inside.get(str(s["id"]), [])}
        for s in sections
    ]


async def replace_sections(
    conn: AsyncConnection,
    action_type_id: UUID,
    sections: list[dict[str, Any]],
) -> None:
    """Write the whole Form tab at once.

    **A whole-document write, like `set_definition` beside it** (decision
    0007). p.124 describes dragging parameters between sections and reordering
    both — a set of granular operations would make the browser send a sequence
    whose intermediate states are forms that do not make sense, and every one
    of them would have to be legal.

    Parameters not named in any section come back to the form body, because
    that is what "moved out of a section" means and there is nowhere else for
    them to go.
    """
    known = {
        str(row["api_name"])
        for row in await fetch_all(
            conn,
            "SELECT api_name FROM action_parameters WHERE action_type_id = :atid",
            {"atid": str(action_type_id)},
        )
    }
    titles: set[str] = set()
    for index, section in enumerate(sections):
        title = str(section.get("title", "")).strip()
        if not title:
            raise ValueError("every section needs a title")
        if title in titles:
            raise ValueError(f"two sections are both called {title!r}")
        titles.add(title)
        for name in section.get("parameters") or []:
            if str(name) not in known:
                raise ValueError(
                    f"{title!r} contains {name!r}, which is not a parameter of "
                    "this action"
                )

    claimed: dict[str, str] = {}
    for section in sections:
        for name in section.get("parameters") or []:
            if str(name) in claimed:
                # p.124 offers two ways to put a parameter in a section and
                # neither is "in both" — a parameter drawn twice would be one
                # box overwriting the other's value as the reader typed.
                raise ValueError(
                    f"{name!r} is in two sections: "
                    f"{claimed[str(name)]!r} and {section.get('title')!r}"
                )
            claimed[str(name)] = str(section.get("title"))

    # **Cleared first, then rebuilt.** `ON DELETE SET NULL` returns every
    # parameter to the form body as the old sections go, so a parameter left
    # out of the new document ends up unsectioned rather than pointing at a row
    # that no longer exists.
    #
    # Everything above runs before this, which reads like the guarantee that a
    # refused form leaves the old one alone — **and is not**. `user_connection`
    # is one transaction per request, so the delete is rolled back whatever
    # order it happened in; a sweep that moved this to the top of the function
    # killed nothing. The order is how the function reads, not what keeps the
    # promise, and the test that checks the promise names the transaction.
    await conn.exec_driver_sql(
        "DELETE FROM action_sections WHERE action_type_id = %s",
        (str(action_type_id),),
    )
    for index, section in enumerate(sections):
        row = await fetch_one(
            conn,
            """
            INSERT INTO action_sections
                   (action_type_id, title, description, columns, collapsible,
                    collapsed, hidden, visible_when, sort_order)
            VALUES (:atid, :title, :description, :columns, :collapsible,
                    :collapsed, :hidden, CAST(:visible AS jsonb), :ord)
            RETURNING id
            """,
            {
                "atid": str(action_type_id),
                "title": str(section["title"]).strip(),
                "description": str(section.get("description") or ""),
                "columns": int(section.get("columns", 1)),
                "collapsible": bool(section.get("collapsible", False)),
                "collapsed": bool(section.get("collapsed", False)),
                "hidden": bool(section.get("hidden", False)),
                "visible": json.dumps(section["visible_when"])
                if section.get("visible_when") else None,
                "ord": index,
            },
        )
        assert row is not None
        names = [str(n) for n in (section.get("parameters") or [])]
        if names:
            await conn.exec_driver_sql(
                "UPDATE action_parameters SET section_id = %s "
                " WHERE action_type_id = %s AND api_name = ANY(%s)",
                (str(row["id"]), str(action_type_id), names),
            )

File: apps/web/src/lib/action-sections.ts
/** Sections on an action form, the parts a document decides (§328;
 * `action-types` p.122-124).
 *
 * > "The action form can be customized with sections. These sections provide a
 * > logical grouping of parameters to organize an action form. Sections also
 * > support columns, descriptions, and conditional overrides." (p.122)
 *
 * > "In the Form tab, click Add section… you can add a title, choose a column
 * > layout, and optionally write a user-facing description. **The description
 * > is not stylized and, unlike parameter descriptions, will always be shown in
 * > the section itself, not in a tooltip.**" (p.123)
 *
 * > "Sections are also collapsible, can be hidden entirely… A section can be
 * > hidden at first and only shown based on a prior parameter." (p.123)
 *
 * > "Parameters and sections display in the form based on their order in this
 * > Form Content section." (p.124)
 *
 * ---
 *
 * **Whether a condition is met is not decided here, and that is the whole
 * point.** p.123's conditional override is decision 0007's `{left, operator,
 * right}` — the same shape a submission criterion uses — and evaluating it in
 * TypeScript would be a second reading of a document the definition editor
 * writes and `check_criteria` runs. `POST .../visible-sections` answers it, and
 * what is in this file is what the *form* decides given that answer: which
 * sections are drawn, which parameters are inside them, which are left in the
 * body, and what to say about the ones that are neither.
 *
 * That is not a shortage of nerve. Reading the condition to learn **which
 * parameters it mentions** is here, in `conditionParameters`, because that is a
 * question about the document's shape rather than about its meaning — and it is
 * what lets a form with plain sections never make the round trip at all.
 *
 * **A section changes what a form looks like and nothing else.** Every function
 * below returns what to draw. A hidden section's parameters are still seeded,
 * still sent and still validated; the server's own module says the same thing
 * in the same words, because a form that quietly dropped values would mean the
 * same action did different things depending on which boxes were on screen.
 */

export interface FormSection {
  id: string;
  title: string;
  description: string;
  columns: number;
  collapsible: boolean;
  collapsed: boolean;
  hidden: boolean;
  /** decision 0007's condition, or `null` for a section that is always on. */
  visible_when: Record<string, unknown> | null;
  /** The parameters inside it, by `api_name`, in the order they are drawn. */
  parameters: string[];
}

/** As much of an action's parameter as a form layout needs. */
export interface FormParameter {
  api_name: string;
  display_name?: string;
  required?: boolean;
  /** p.25's hidden parameter: supplied by the caller and never drawn, whatever
   * any section says about it. */
  hidden?: boolean;
}

function side(spec: unknown): Record<string, unknown> {
  return spec && typeof spec === "object" && !Array.isArray(spec)
    ? (spec as Record<string, unknown>)
    : {};
}

/** Every parameter any section's condition reads.
 *
 * **The reason a plain form never asks the server anything.** A section with no
 * condition cannot change its mind, and one that is hidden entirely is not
 * asking a question — so if this comes back empty there is nothing to watch and
 * nothing to send.
 *
 * Reading `left` and `right` for `kind: "parameter"` is not evaluating the
 * condition: an operator this build has never heard of still names its
 * parameters here, and a condition that mentions the current user rather than a
 * parameter contributes nothing — which is correct, because who is asking does
 * not change while a form is open.
 */
export function conditionParameters(sections: FormSection[]): string[] {
  const named = new Set<string>();
  for (const section of sections ?? []) {
    if (section.hidden) continue;
    const condition = side(section.visible_when);
    for (const key of ["left", "right"]) {
      const spec = side(condition[key]);
      if (spec.kind !== "parameter") continue;
      const name = String(spec.parameter ?? "").trim();
      if (name) named.add(name);
    }
  }
  return [...named].sort();
}

/** Whether any section's visibility is a question at all.
 *
 * **This, not `conditionParameters`, is what decides whether to ask the
 * server** — and the first draft used the other one, which was a real defect
 * and the reason this function exists. p.50 gives two condition templates, and
 * only one of them names a parameter: a section shown to the person a
 * criterion names ("based on current user") reads nothing out of the form at
 * all, so `conditionParameters` came back empty, the form decided it had
 * nothing to ask about, and that section could never be drawn for anybody.
 *
 * The two questions are genuinely different and were folded into one:
 * *whether* the server has to be asked (this) and *when the answer goes stale*
 * (`conditionKey`). A condition about the current user is asked once and never
 * again, because who is asking does not change while a form is open.
 */
export function hasConditions(sections: FormSection[]): boolean {
  return (sections ?? []).some((s) => !s.hidden && !!s.visible_when);
}

/** A stable key over only the values the conditions read.
 *
 * What the form's query is keyed on, so typing in a parameter no section asks
 * about does not re-ask the server — and so the answer for values already seen
 * is the cached one rather than a round trip and a flicker.
 *
 * **A constant when no condition names a parameter**, which is right rather
 * than a degenerate case: a section shown to one person (p.50's "based on
 * current user") has an answer that cannot change while the form is open, so
 * one key means one round trip. Whether to make that round trip is
 * `hasConditions`, not this — see the note there.
 */
export function conditionKey(
  sections: FormSection[],
  values: Record<string, unknown>,
): string {
  const named = conditionParameters(sections);
  return JSON.stringify(named.map((name) => [name, values?.[name] ?? null]));
}

/** Which sections are drawn, given the server's answer.
 *
 * p.123's three states and one more the page has that the server does not:
 *
 * * `hidden` — "hidden entirely". Out, whatever else says.
 * * no condition — always in. Every section made before anybody configured one.
 * * a condition, and an answer in hand — in when the answer names it.
 * * a condition, and **no answer yet** — out.
 *
 * The last is p.123's own wording: a section is "hidden at first and only shown
 * based on a prior parameter". Drawing it while the question is in flight would
 * put the boxes on screen for a moment and then take them away, which is worse
 * than the wait — and it is the same direction the server fails in when it
 * cannot read a condition at all.
 */
export function drawnSections(
  sections: FormSection[],
  visible: string[] | undefined,
): FormSection[] {
  const shown = visible ? new Set(visible) : null;
  return (sections ?? []).filter((section) => {
    if (section.hidden) return false;
    if (!section.visible_when) return true;
    return shown !== null && shown.has(section.id);
  });
}

export interface DrawnSection {
  section: FormSection;
  parameters: FormParameter[];
}

export interface FormLayout {
  /** Parameters in no section at all, drawn in the form body. */
  loose: FormParameter[];
  /** p.124's Form Content order, filtered to what is on screen. */
  sections: DrawnSection[];
}

/** p.124's form, arranged.
 *
 * **The body first, then the sections.** p.124 says "Parameters and sections
 * display in the form based on their order in this Form Content section" — one
 * order over both — and this build stores two: a section's `sort_order` and a
 * parameter's. A parameter that has never been put in a section keeps its place
 * among the parameters, and the sections follow in theirs. Interleaving them
 * needs a single ordering across both kinds, which is a change to the document
 * rather than to this function, and `docs/parity` carries it as its own row.
 *
 * A parameter named by a section is drawn **only** in that section: absent from
 * the body when the section is on screen, and absent from the form entirely
 * when it is not. Falling back to the body would undo p.123's hiding by the
 * most direct route available.
 */
export function formLayout(
  parameters: FormParameter[],
  sections: FormSection[],
  visible: string[] | undefined,
): FormLayout {
  // p.25 first, and over every parameter rather than only the loose ones: a
  // hidden parameter is supplied by whatever runs the action, and putting one
  // in a section does not make it something to fill in.
  const drawable = (parameters ?? []).filter((p) => !p.hidden);
  const byName = new Map(drawable.map((p) => [p.api_name, p]));
  // Claimed by *any* section, drawn or not.
  const claimed = new Set(
    (sections ?? []).flatMap((s) => s.parameters ?? []),
  );
  return {
    loose: drawable.filter((p) => !claimed.has(p.api_name)),
    sections: drawnSections(sections, visible).map((section) => ({
      section,
      parameters: (section.parameters ?? [])
        .map((name) => byName.get(name))
        .filter((p): p is FormParameter => !!p),
    })),
  };
}

/** Required parameters the form is not showing anywhere.
 *
 * The case p.123 does not discuss and a builder will reach on their second
 * afternoon: a required parameter inside a section that is hidden, or whose
 * condition is not met. The server still requires it, so the submission is
 * refused — and a form that said "Priority is required" beside no Priority box
 * is §214's shape, a control that looks like it works.
 *
 * So the form keeps the button disabled, which is true, and says something a
 * reader can act on: see `unreachableNote`.
 */
export function requiredElsewhere(
  parameters: FormParameter[],
  sections: FormSection[],
  visible: string[] | undefined,
): FormParameter[] {
  const layout = formLayout(parameters, sections, visible);
  const onScreen = new Set([
    ...layout.loose.map((p) => p.api_name),
    ...layout.sections.flatMap((d) => d.parameters.map((p) => p.api_name)),
  ]);
  return (parameters ?? []).filter(
    (p) => !p.hidden && p.required && !onScreen.has(p.api_name),
  );
}

export function labelOf(parameter: FormParameter): string {
  return parameter.display_name || parameter.api_name;
}

/** What to say about `requiredElsewhere`, or `null` when there is nothing.
 *
 * Addressed to whoever can fix it rather than to whoever is stuck: the reader
 * of a form cannot open a section somebody hid, and telling them to fill in a
 * field would be an instruction they cannot follow.
 */
export function unreachableNote(missing: FormParameter[]): string | null {
  if (!missing || missing.length === 0) return null;
  const names = missing.map(labelOf).join(", ");
  return missing.length === 1
    ? `${names} is required but is in a section this form does not show. ` +
      "Whoever arranged the form has to move it or show the section."
    : `${names} are required but are in sections this form does not show. ` +
      "Whoever arranged the form has to move them or show the sections.";
}

/** p.123: "A section can be divided into one or two columns." */
export const COLUMN_CHOICES = [1, 2];

export function columnsOf(section: { columns?: unknown }): number {
  return section?.columns === 2 ? 2 : 1;
}

export function columnsLabel(columns: unknown): string {
  return columns === 2 ? "Two columns" : "One column";
}

/** Whether a section starts folded.
 *
 * **`collapsed` means nothing without `collapsible`**, which the migration says
 * in the column's own comment: a section folded with no way to open it would be
 * p.123's "hidden entirely" wearing the wrong name, and a builder who ticked
 * one box and not the other would have made a form whose fields cannot be
 * reached without knowing that.
 */
export function collapsedInitially(section: FormSection): boolean {
  return !!section?.collapsible && !!section?.collapsed;
}

/** One line about a section, for p.124's Form tab list. */
export function sectionSummary(section: FormSection): string {
  const count = (section?.parameters ?? []).length;
  const parts = [
    count === 1 ? "1 parameter" : `${count} parameters`,
    columnsLabel(section?.columns).toLowerCase(),
  ];
  if (section?.hidden) parts.push("hidden");
  else if (section?.visible_when) parts.push("shown on a condition");
  if (section?.collapsible) {
    parts.push(collapsedInitially(section) ? "collapsible, starts folded" : "collapsible");
  }
  return parts.join(" · ");
}

/** The parameters a given section may still be given.
 *
 * **Narrowing what is declarable, not validating.** The server refuses a
 * parameter that is in two sections, and this is why nobody reaches that
 * refusal by the obvious route: the dropdown for section *n* does not offer the
 * ones section *m* already holds. A parameter this section holds is still
 * offered — removing it from its own list would make the control forget what it
 * is showing.
 */
export function availableParameters(
  parameters: FormParameter[],
  sections: FormSection[],
  index: number,
): FormParameter[] {
  const elsewhere = new Set(
    (sections ?? []).flatMap((s, i) => (i === index ? [] : s.parameters ?? [])),
  );
  return (parameters ?? []).filter((p) => !elsewhere.has(p.api_name));
}

/** Put a parameter in one section, taking it out of wherever it was.
 *
 * p.124 offers two ways to put a parameter in a section and neither is "in
 * both". A move is one edit as far as the builder is concerned, so it is one
 * here too — the alternative is a Form tab that can sit in a state the server
 * refuses until somebody notices which other section still claims the name.
 */
export function placeParameter(
  sections: FormSection[],
  index: number,
  name: string,
): FormSection[] {
  return (sections ?? []).map((section, i) => {
    const without = (section.parameters ?? []).filter((p) => p !== name);
    return i === index
      ? { ...section, parameters: [...without, name] }
      : { ...section, parameters: without };
  });
}

export function removeParameter(
  sections: FormSection[],
  index: number,
  name: string,
): FormSection[] {
  return (sections ?? []).map((section, i) =>
    i === index
      ? { ...section, parameters: (section.parameters ?? []).filter((p) => p !== name) }
      : section,
  );
}

/** Move a section up or down p.124's Form Content list.
 *
 * A no-op at the ends rather than a wrap: a list that jumped from top to bottom
 * because somebody pressed the arrow once too often would be a reorder nobody
 * asked for, and the button is disabled there anyway.
 */
export function moveSection(
  sections: FormSection[],
  index: number,
  delta: number,
): FormSection[] {
  const next = [...(sections ?? [])];
  const to = index + delta;
  if (index < 0 || index >= next.length || to < 0 || to >= next.length) return next;
  const moved = next.splice(index, 1);
  next.splice(to, 0, ...moved);
  return next;
}

/** A blank section, as Add section leaves it (p.123). */
export function blankSection(existing: FormSection[]): FormSection {
  const taken = new Set((existing ?? []).map((s) => s.title));
  let title = "Section";
  for (let n = 2; taken.has(title); n += 1) title = `Section ${n}`;
  return {
    id: "",
    title,
    description: "",
    columns: 1,
    collapsible: false,
    collapsed: false,
    hidden: false,
    visible_when: null,
    parameters: [],
  };
}

/** p.123's "shown based on a prior parameter", as the Form tab edits it.
 *
 * The stored shape is decision 0007's `{left, operator, right}` — general
 * enough to compare two parameters or ask about the current user — and the
 * editor offers the one case p.123 actually describes: a prior parameter
 * against a value. **Read back tolerantly**: a condition written by hand or by
 * an import may be something this control cannot express, and `expressible`
 * says so rather than the editor quietly rewriting it into something it can.
 */
export interface ConditionDraft {
  /** The prior parameter, or `""` for a section with no condition. */
  parameter: string;
  operator: string;
  value: string;
  /** Whether the control can show what is stored without changing it. */
  expressible: boolean;
}

export function conditionDraft(section: FormSection): ConditionDraft {
  const condition = side(section?.visible_when);
  if (!section?.visible_when || Object.keys(condition).length === 0) {
    return { parameter: "", operator: "is", value: "", expressible: true };
  }
  const left = side(condition.left);
  const right = side(condition.right);
  const parameter = left.kind === "parameter" ? String(left.parameter ?? "") : "";
  const value = right.kind === "value" ? String(right.value ?? "") : "";
  return {
    parameter,
    operator: String(condition.operator ?? "is"),
    value,
    // A condition comparing two parameters, or asking about the current user,
    // is legal and the server evaluates it — this control just cannot draw it,
    // and a dropdown that silently reduced it to "no condition" would delete a
    // rule on the next save.
    expressible: !!parameter && right.kind === "value",
  };
}

/** The condition a draft stores, or `null` for "always shown".
 *
 * An empty parameter is not a condition: p.123's conditional override names a
 * prior parameter, and a half-filled row in a settings panel is what somebody
 * who changed their mind leaves behind. `null` and a condition that happens to
 * be true are different things — only the first survives the parameter being
 * renamed.
 */
export function conditionValue(draft: ConditionDraft): Record<string, unknown> | null {
  if (!draft?.parameter?.trim()) return null;
  return {
    left: { kind: "parameter", parameter: draft.parameter.trim() },
    operator: draft.operator || "is",
    right: { kind: "value", value: draft.value },
  };
}

/** Carry a parameter rename through the form.
 *
 * The same edit `patchParameter` makes to rules and criteria, for the same
 * reason: a section still naming the old parameter makes the server refuse the
 * save, and the refusal is about a row the person did not touch. A rename to
 * nothing is not a rename — an api_name is cleared mid-typing on the way to the
 * next one, and taking the parameter out of its section at that moment would
 * lose the arrangement for a keystroke.
 */
export function renameParameter(
  sections: FormSection[],
  before: string,
  after: string,
): FormSection[] {
  if (!before || !after || before === after) return sections ?? [];
  return (sections ?? []).map((section) => {
    const condition = conditionDraft(section);
    const renamed = condition.expressible && condition.parameter === before
      ? conditionValue({ ...condition, parameter: after })
      : section.visible_when;
    return {
      ...section,
      parameters: (section.parameters ?? []).map((p) => (p === before ? after : p)),
      visible_when: renamed,
    };
  });
}

/** Take a parameter the action no longer declares out of every section.
 *
 * Deleting a parameter is the other edit with consequences here. The server
 * would refuse the form for naming something that is not a parameter — true,
 * and about a section the person was not looking at.
 */
export function forgetParameters(
  sections: FormSection[],
  declared: FormParameter[],
): FormSection[] {
  const known = new Set((declared ?? []).map((p) => p.api_name));
  return (sections ?? []).map((section) => ({
    ...section,
    parameters: (section.parameters ?? []).filter((name) => known.has(name)),
  }));
}
